packages/costa/src/runnable.ts

```diff
@@ -244,22 +276,42 @@ export class PluginRunnable {
    * @param op
    * @param msg
    */
-  private async sendAndWait(op: PluginOperator, msg: PluginMessage): Promise<PluginMessage> {
-    await this.send(op, msg);
-    debug(`sent ${msg.event} for ${this.id}, and wait for response`);
+  private async sendAndWait(op: PluginOperator, msg: PluginMessage, timeout?: number): Promise<PluginMessage> {
+    return new Promise<PluginMessage>((resolve, reject) => {
+      let opNotRespondingTimer: NodeJS.Timer;
+      if (typeof timeout === 'number' && timeout > 0) {
+        opNotRespondingTimer = setTimeout(() => {
```

Comment:
I think we could insert the timer into function `send` like this:
```js
  /**
   * Reads the message, it's blocking the async context util.
   * @params timeout timeout value in millisecond
   */
  private async read(timeout = 0): Promise<PluginProtocol> {
    return new Promise((resolve, reject) => {
     let timer;
      if (timeout > 0) {
        timer = setTimeout(() => {
          reject(new Error('read timeout'));
        }, timeout);
      }
      this.onread = (res: PluginProtocol) => {
        if (timer) {
          clearTimeout(timer);
        }
        resolve(res);
      };
      this.onreadfail = (err: Error) => {
        if (timer) {
          clearTimeout(timer);
        }
        reject(err);
      };
    });
  }
```
Is this a little more concise?